Replace vendored `libmad` with submodule of tenacityteam's `libmad`.

sukibaby opened the pull request:
- [x]  Integrate libmad 0.16 as submodule
- [x]  Compiles & runs as expected with slightly tweaked automake configuration
- [ ]  Fully replace existing libmad build system with libmad 0.16's CMake implementation

sukibaby committed:
Remove vendored libmad directory

sukibaby committed:
Bring in Tenacity team's libmad fork as submodule

sukibaby committed:
Prepare for libmad 0.16+

Revise libmad CMake configuration (source tree restructuring, reorganize order of operations)
Remove unneeded or outdated libmad configuration
Define best options for modern 64-bit CPU's and OS's
Rearrange cmake configuration logic for easier reading

Notes:
OPT_SPEED should be preferred to OPT_ACCURACY, because OPT_ACCURACY could cause stutters on corrupted/damaged/non-compliant MP3 files.
When profiling ASO_INTERLEAVE1 vs ASO_INTERLEAVE2, ASO_INTERLEAVE2 was faster on 64-bit CPU compiled with MSVC.
The generic 64-bit FPM is used to maximize portability, but also any hardware capable of running ITGMania can handle whatever FPM we use.
I removed the line disabling project warnings because we don't expect any in libmad 0.16+.

sukibaby commented:
closing while i try something out in CI to fix the mac build here

sukibaby commented:
all good now - had to include `stdlib.h` for Mac - squashing

teejusb commented on .gitmodules:
```diff
@@ -50,3 +50,6 @@
 	shallow = true
+[submodule "extern/libmad"]
+	path = extern/libmad
+	url = https://codeberg.org/tenacityteam/libmad.git
```
Can we mark it as shallow

sukibaby replied:
sure

teejusb commented on src/RageSoundReader_MP3.cpp:
```diff
@@ -22,7 +23,7 @@ enum tagtype {
 static const int ID3_TAG_FLAG_FOOTERPRESENT = 0x10;
 
-static tagtype tagtype( const unsigned char *data, id3_length_t length )
+static tagtype detect_tagtype( const unsigned char *data, id3_length_t length )
```
Why is this deemed necessary? Seems just like a name change.

sukibaby replied:
Not really necessary. It will still build without this change. This change suppresses two errors which were not present with the existing vendored libmad

E0757 function "tagtype" is not a type name
E0311 cannot overload functions distinguished by return type alone

 So I can undo it.

teejusb replied:
AH I see it's a naming conflict with the return type. Can keep your change.

sukibaby committed:
RageSoundReader_MP3: resolve naming conflict

Because of a naming collision with the return type, MSVC emits errors E0757 (function "tagtype" is not a type name) & E0311 (cannot overload functions distinguished by return type alone). This is resolved by renaming the function itself so that there is not a naming conflict.

sukibaby commented:
~~I need to do some testing to see why CodeQL isn't finding the `config.h` generated by cmake, rather than the pre-packaged `config.mad.in.h` in the extern directory. CI is passing but CodeQL isn't when I point to our generated `config.h`  instead of the premade one.~~

Resolved.

sukibaby commented:
Did some Ubuntu testing, should be good now.

edit::  Yup all good, Linux/CodeQL aren't complaining of missing `config.h` now. will squash

sukibaby committed:
Provide better context in libmad cmake comments.

teejusb commented:
I'm going to hold off on this since looking the commits on the new version don't really look particularly substantial and I'd rather wait to do a wholesale swap of the lib as it's not pressing for this release.

sukibaby commented:
Np, I will close until after the new version  to reduce clutter.

sukibaby commented:
reopening since new version is out
